spot_rl_experiments/spot_rl/envs/place_env.py

```diff
@@ -2,69 +2,249 @@
-    env = SpotPlaceEnv(config, spot, place_target, args.target_is_local)
-    env.power_robot()
-    policy = PlacePolicy(config.WEIGHTS.PLACE, device=config.DEVICE)
-    policy.reset()
-    observations = env.reset()
-    done = False
-    env.say("Starting episode")
-    while not done:
-        action = policy.act(observations)
-        observations, _, done, _ = env.step(arm_action=action)
-    if done:
-        while True:
-            env.reset()
-            spot.set_base_velocity(0, 0, 0, 1.0)
+    return config
+
+
+class PlaceController:
+    """
+    Place controller is used to execute place for given place targets
+
+    Args:
+        config: Config object
+        spot: Spot object
+        use_policies (bool): Whether to use policies or use BD API to execute place
+
+    How to use:
+        1. Create PlaceController object
+        2. Call execute() with place_target_list as input
+        3. Call shutdown() to stop the robot
+
+    Example:
+        config = construct_config_for_place(opts=[])
+        spot = Spot("PlaceController")
+        place_target_list = [target1, target2, ...]
+        place_controller = PlaceController(config, spot, use_policies=True)
+        place_result = place_controller.execute(place_target_list)
+        place_controller.shutdown()
+    """
+
+    def __init__(self, config, spot: Spot, use_policies=True):
+        self.spot = spot
+        self.use_policies = use_policies
+
+        # Setup
+        self.policy = PlacePolicy(config.WEIGHTS.PLACE, device=config.DEVICE)
+        self.policy.reset()
+
+        self.place_env = SpotPlaceEnv(config, spot)
+        self.place_env.power_robot()
+
+    def execute(self, place_target_list, is_local=False):
+        """
+        Execute place for each place target in place_target_list
+
+        Args:
+            place_target_list (list): List of place targets to go and place
+            is_local (bool): Whether the place target is in the local frame of the robot
+
+        Returns:
+            success_list (list): List of dicts containing the following keys:
+                - time_taken (float): Time taken to place the object
+                - success (bool): Whether the place was successful
+                - place_target (np.array([x,y,z])): Place target in base frame
+                - ee_pos (np.array([x,y,z])): End effector position in base frame
+        """
+        success_list = []
+        for place_target in place_target_list:
+            start_time = time.time()
+            if self.use_policies:
+                observations = self.place_env.reset(place_target, is_local)
+                done = False
+
+                while not done:
+                    action = self.policy.act(observations)
+                    observations, _, done, _ = self.place_env.step(arm_action=action)
+
+            else:
+                # Get reset arm position (Is there a better way to do this without using place_env?????????????)
```

Comment:
Any suggestions here @jimmytyyang  @akshararai

Reply:
Sorry, do you mean how to reset the arm joint?

Reply:
yes

Reply:
This can be tracked by https://github.com/facebookresearch/spot-sim2real/issues/67